### PR TITLE
Add auto pairs for same-char pairs

### LINKED ISSUE #1014: add support for autopairs of certain keys
#### Describe your feature request

allow autopairs of ```"``` and ```'``` keys
personally i think the first one: ```"``` is most commonly used for strings, and should thus be autopaired.
second one: ```'``` can sometimes be used in lifetime annotations i guess, so not necessary to add this one.

### PR DESCRIPTION
**Pull request by dead10ck**
## Changes

* Add unit tests for all existing functionality
* Add auto pairs for same-char pairs (quotes, etc). Account for apostrophe in prose by requiring both sides of the cursor to be non-pair chars or whitespace
* Slight factor of moving the cursor transform of the selection to inside the hooks. This will enable doing auto pairing with selections, and fixing the bug where auto pairs destroy the selection.

I will still be working on other things, but I thought it would be a good idea to make an incremental smaller PR for the sake of ease of review, and for providing opportunity for feedback with the approach to testing and implementation details before I build further on them.

Fixes #1014

Things I'd still like to do:

* delete implementation where it erases the whole bracket (|) -> |
* do not reduce to cursors; use whole selections, and surround with pair
* change to multi character pairs to handle cases like placing the cursor in the middle of triple quotes, and more exotic pairs like Jinja's `{% %}`
* add more granular config for particular pairs and behaviors as discussed in #992

In particular about the auto pairing with selections, I think when we have a case like (where `[]` is the selection)

```
[wor]d
```

and we enter insert mode and type `"`, we could either unconditionally surround the selection and make

```
"[wor]"d
```

so it behaves exactly like surround, or we could choose only to surround when both sides of the selection are non-pair chars or whitespace, extending the current behavior to selections, so

```
"[wor]d
```

I'm on the fence about it, so any guidance there would be helpful.

### CODE REVIEW

File: helix-core/src/auto_pairs.rs [hunk 1]
```diff
@@ -100,64 +115,337 @@ fn handle_open(
...
+        test_hooks_with_pairs(
+            &Rope::from("word"),
+            &Selection::single(1, 0),
+            PAIRS,
+            |open, _| format!("{}word", open),
+            &Selection::point(1),
+        )
+    }
+
+    // [TODO] broken until it works with selections
```
Comment (archseer) [inline helix-core/src/auto_pairs.rs:409]: This is sort of addressed byb`msa"`(match surrounding add `(`)

File: helix-core/src/auto_pairs.rs [hunk 2]
```diff
@@ -100,64 +115,337 @@ fn handle_open(
...
-    // if not cursor but selection, wrap
-    // let next = next char
-
-    // if next == bracket {
-    //   // if start of syntax node, insert token twice (new pair because node is complete)
-    //   // elseif colsedBracketAt
-    //      // is_triple == allow triple && next 3 is equal
-    //      // cursor jump over
-    // }
-    //} else if allow_triple && followed by triple {
```
Comment (archseer) [inline helix-core/src/auto_pairs.rs:156]: The proposed algorithm here from codemirror also handled triples correctly (for docstrings for example), can you implement it/add a TODO?

**Comment by archseer** [comment]
Great work! I need to give this a full read later but it's especially great to see test coverage being added :)

File: helix-core/src/auto_pairs.rs (hunk 1 above)
Comment (dead10ck) [reply helix-core/src/auto_pairs.rs:409]: Sorry, just to be clear, are you saying we shouldn't worry about adding surround functionality with selections? Or just that this particular case should just be inserting a ( before the word, but if the whole word was selected, do the surround?

File: helix-core/src/auto_pairs.rs (hunk 2 above)
Comment (dead10ck) [reply helix-core/src/auto_pairs.rs:156]: The CodeMirror code for this looks pretty hand tailored specifically for triplets, and I'm planning to cover this when I change the open/close pairs to be arbitrary strings, so it can place the cursor in the middle for n-char pairs. There's a TODO for it near the top.

File: helix-core/src/auto_pairs.rs (hunk 1 above)
Comment (dead10ck) [reply helix-core/src/auto_pairs.rs:409]: Ok, I've fixed this test, although it's still broken until I add the changes for selections.

File: helix-core/src/auto_pairs.rs (hunk 2 above)
Comment (archseer) [reply helix-core/src/auto_pairs.rs:156]: It deals with triplets specially so that it can handle `""|`  -> `"""|"""`, so that it understands the " is a part of a triplet and not just closing a pair.

Comment (dead10ck) [reply helix-core/src/auto_pairs.rs:156]: Yep. And when pairs are arbitrary strings, it will have a similar understanding. It will additionally be able to handle `{|}` → `{%|%}`, `f'|` → `f'|'`, `ba|` → `bar|quux` or anything else. Handling triplets specially would be redundant.

**Review by archseer — CHANGES_REQUESTED** [review]

File: helix-core/src/auto_pairs.rs [hunk 3]
```diff
@@ -25,71 +28,83 @@ const CLOSE_BEFORE: &str = ")]}'\":;> \n\r\u{000B}\u{000C}\u{0085}\u{2028}\u{202
...
+
 fn handle_open(
     doc: &Rope,
     selection: &Selection,
     open: char,
     close: char,
     close_before: &str,
 ) -> Transaction {
-    let mut ranges = SmallVec::with_capacity(selection.len());
+    let mut end_ranges = SmallVec::with_capacity(selection.len());
```
[the remark this answers is not in the document]
Comment (dead10ck) [reply helix-core/src/auto_pairs.rs:78]: Yeah, sorry, this was mostly for my own benefit to better understand the code. When building the final range, in the anchor position, it's offset inline, but the head is in a variable that was defined above. And to me `pos` (i.e. position) and `head` are kind of synonymous, and it was hard to follow that one is what we started with and the other is what we want it to be after the text transformation. Are you okay with the changes?

File: helix-core/src/auto_pairs.rs [hunk 4]
```diff
@@ -25,71 +28,83 @@ const CLOSE_BEFORE: &str = ")]}'\":;> \n\r\u{000B}\u{000C}\u{0085}\u{2028}\u{202
...
 
-// TODO: special handling for lifetimes in rust: if preceeded by & or < don't auto close '
-// for example "&'a mut", or "fn<'a>"
-
 fn next_char(doc: &Rope, pos: usize) -> Option<char> {
     if pos >= doc.len_chars() {
         return None;
     }
     Some(doc.char(pos))
 }
```
[the remark this answers is not in the document]
Comment (dead10ck) [reply helix-core/src/auto_pairs.rs]: Right, done.

File: helix-core/src/auto_pairs.rs [hunk 5]
```diff
@@ -100,64 +115,337 @@ fn handle_open(
...
+        let start_head = start_range.head;
+        let end_head = start_head + offs + token.len_utf8();
+
+        let end_anchor = if start_range.is_empty() {
+            end_head
+        } else {
+            start_range.anchor + offs
+        };
+
+        // if selection, retain anchor, if cursor, move over
```
[the remark this answers is not in the document]
Comment (dead10ck) [reply helix-core/src/auto_pairs.rs]: Done

**Review by archseer — APPROVED** [review]
Looks good to me but I'd like to get one more reviewer

/cc @sudormrfbin

**Review by sudormrfbin — COMMENTED** [review]
Some minor nitpicks, other than that LGTM :)

File: helix-core/src/auto_pairs.rs [hunk 6]
```diff
@@ -25,71 +28,76 @@ const CLOSE_BEFORE: &str = ")]}'\":;> \n\r\u{000B}\u{000C}\u{0085}\u{2028}\u{202
...
-// TODO: delete implementation where it erases the whole bracket (|) -> |
+// [TODO]
+// * delete implementation where it erases the whole bracket (|) -> |
+// * do not reduce to cursors; use whole selections, and surround with pair
+// * change to multi character pairs to handle cases like placing the cursor in the
+//   middle of triple quotes, and more exotic pairs like Jinja's {% %}
 
 #[must_use]
 pub fn hook(doc: &Rope, selection: &Selection, ch: char) -> Option<Transaction> {
+    debug!("autopairs hook selection: {:#?}", selection);
```
[the remark this answers is not in the document]
Comment (dead10ck) [reply helix-core/src/auto_pairs.rs:39]: Not stray per se. I found it very useful for debugging, so I thought it made sense to leave in there for debug logs. Do you think it would be better as a trace? Or do you think it's too noisy to stay?

File: helix-core/src/auto_pairs.rs [hunk 7]
```diff
@@ -25,71 +28,76 @@ const CLOSE_BEFORE: &str = ")]}'\":;> \n\r\u{000B}\u{000C}\u{0085}\u{2028}\u{202
 //
 // to simplify, maybe return Option<Transaction> and just reimplement the default
 
-// TODO: delete implementation where it erases the whole bracket (|) -> |
+// [TODO]
+// * delete implementation where it erases the whole bracket (|) -> |
+// * do not reduce to cursors; use whole selections, and surround with pair
```
[the remark this answers is not in the document]
Comment (dead10ck) [reply helix-core/src/auto_pairs.rs:33]: It will be once https://github.com/helix-editor/helix/pull/1254 is merged, but that is written on top of this PR.

File: helix-core/src/auto_pairs.rs [hunk 8]
```diff
@@ -25,71 +28,76 @@ const CLOSE_BEFORE: &str = ")]}'\":;> \n\r\u{000B}\u{000C}\u{0085}\u{2028}\u{202
...
+            end_head
+        } else {
+            start_range.anchor + offs
+        };
+
+        end_ranges.push(Range::new(end_anchor, end_head));
 
         match next {
             Some(ch) if !close_before.contains(ch) => {
                 offs += 1;
```
[the remark this answers is not in the document]
Comment (dead10ck) [reply helix-core/src/auto_pairs.rs:98]: Good catch
